Transfer generic stuff from eager to template (#788)
nf-core/tools: nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/CONTRIBUTING.md

* Fix the bug, and bump version (X.Y.Z+1).
* A PR should be made on `master` from patch to directly this particular bug.

> ## Getting Help

Reviewer: I think the vast majority of nf-core docs only capitalise the first word of headings. So I'd prefer to change the others and leave this one as it is given the choice.

For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://nf-co.re/{{ cookiecutter.short_name }}/usage) and don't hesitate to get in touch on the nf-core Slack [#{{ cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).

## Pipeline Contribution Conventions

To make the {{ cookiecutter.short_name }} code and processing logic more understandable for new contributors and to ensure quality, we semi-standardise the way the code and other contributions are written.

### Adding a New Module

> If you wish to contribute a new module, please use the following coding standards:

Reviewer: I think the use of the word "module" here is a bit confusing, given DSL2. I guess you just mean a new step in the pipeline?

2. Write the process block (see below).
3. Define the output channel if needed (see below).
4. Add any new flags/options to `nextflow.config` with a default (see below).
5. Add any new flags/options to `nextflow_schema.json` with help text (with `nf-core schema build .`)
6. Add any new flags/options to the help message (for integer/text parameters, print to help the corresponding `nextflow.config` parameter).
7. Add sanity checks for all relevant parameters.
8. Add any new software to the `scrape_software_versions.py` script in `bin/` and the version command to the `scrape_software_versions` process in `main.nf`.
9. Do local tests that the new code works properly and as expected.
10. Add a new test command in `.github/workflow/ci.yaml`.
11. If applicable add a [MultiQC](https://https://multiqc.info/) module.
12. Update MultiQC config `assets/multiqc_config.yaml` so relevant suffixes, name clean up, General Statistics Table column order, and module figures are in the right order.
13. Optional: Add any descriptions of MultiQC report sections and output files to `docs/output.md`.

> 
> More details are as follow

Reviewer: ```suggestion
```

### Default Values

> Default values should go in `nextflow.config` under the `params` scope, and `nextflow_schema.json` (latter with `nf-core schema build .`)

Reviewer: ```suggestion
Parameters should be initialised / defined with default values in `nextflow.config` under the `params` scope. 
Once there, use `nf-core schema build .` to add to `nextflow_schema.json`.
```

### Default Processes Resource Requirements

> Defining recommended 'minimum' resource requirements (CPUs/Memory) for a process should be defined in `conf/base.config`. This can be utilised within the process using `${task.cpu}` or `${task.memory}` variables in the `script:` block.

Reviewer: ```suggestion
Defining sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. This can be utilised within the process using `${task.cpu}` or `${task.memory}` variables in the `script:` block.
```

Reviewer: Would also be good to add mention of tagging processes with labels and trying to stick to the standard nf-core label names where possible (see eg. [the rnaseq pipeline](https://github.com/nf-core/rnaseq/blob/bc5fc76f40b2da6082a854927184c9d6e5060393/conf/base.config#L23-L40)).

Author: Hmm, I actually wanted to ask about the latter - I saw that rnaseq is apparently the 'standard', however it doesn't account for single CPU processes... is this something I should bring up on slack for wider discussion?

In eager I have basically what is in rnaseq, but with two versions: `sc` for single core and `mc` for multicore.

Reviewer: The base config (ie no label) can be used for single CPU processes: https://github.com/nf-core/rnaseq/blob/bc5fc76f40b2da6082a854927184c9d6e5060393/conf/base.config#L14-L16

Author: Hmm OK. And I guess you do process specific increases for when they need more memory?

### Naming Schemes

Please use the following naming schemes, to make it easy to understand what is going where.

* initial process channel: `ch_output_from_<process>`
* intermediate and terminal channels: `ch_<previousprocess>_for_<nextprocess>`

### Nextflow Version Bumping

> If you have agreement from reviewers, you may bump the 'default' minimum version of nextflow (e.g. for testing), with `nf-core bump-version`.

Reviewer: ```suggestion
If you have agreement from reviewers, you may bump the 'default' minimum version of nextflow (e.g. for testing), with: `nf-core bump-version --nextflow . [min-nf-version]`
```

### Software Version Reporting

If you add a new tool to the pipeline, please ensure you add the information of the tool to the `get_software_version` process.

Add to the script block of the process, something like the following:

```bash
<YOUR_TOOL> --version &> v_<YOUR_TOOL>.txt 2>&1 || true
```

or

```bash
<YOUR_TOOL> --help | head -n 1 &> v_<YOUR_TOOL>.txt 2>&1 || true
```

You then need to edit the script `bin/scrape_software_versions.py` to

1. add a (python) regex for your tools --version output (as in stored in the `v_<YOUR_TOOL>.txt` file), to ensure the version is reported as a `v` and the version number e.g. `v2.1.1`
2. add a HTML block entry to the `OrderedDict` for formatting in MultiQC.

> > If a tool does not unfortunately offer any printing of version data, you may add this 'manually' e.g. with `echo "v1.1" > v_<YOUR_TOOL>.txt`

Reviewer: eek, I'd rather leave out this comment. Hopefully this is so rare as to be almost non-existent and I can imagine it going very wrong.

## Getting help

For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://nf-co.re/{{ cookiecutter.short_name }}/usage) and don't hesitate to get in touch on the nf-core Slack [#{{ cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).

## Pipeline contribution conventions

> To make the {{ cookiecutter.short_name }} code and processing logic more understandable for new contributors and to ensure quality, we semi-standardise the way the code and other contributions are written.

Reviewer: ```suggestion
To make the {{ cookiecutter.name }} code and processing logic more understandable for new contributors and to ensure quality, we semi-standardise the way the code and other contributions are written.
```

5. Add any new flags/options to `nextflow_schema.json` with help text (with `nf-core schema build .`)
6. Add any new flags/options to the help message (for integer/text parameters, print to help the corresponding `nextflow.config` parameter).
7. Add sanity checks for all relevant parameters.
8. Add any new software to the `scrape_software_versions.py` script in `bin/` and the version command to the `scrape_software_versions` process in `main.nf`.
9. Do local tests that the new code works properly and as expected.
10. Add a new test command in `.github/workflow/ci.yaml`.
11. If applicable add a [MultiQC](https://https://multiqc.info/) module.
12. Update MultiQC config `assets/multiqc_config.yaml` so relevant suffixes, name clean up, General Statistics Table column order, and module figures are in the right order.
13. Optional: Add any descriptions of MultiQC report sections and output files to `docs/output.md`.

### Default values

Parameters should be initialised / defined with default values in `nextflow.config` under the `params` scope.

Once there, use `nf-core schema build .` to add to `nextflow_schema.json`.

### Default processes resource requirements

> Sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. These should generally be specified generic with `withLabel:` selectors so they can be shared across multiple processes/steps of the pipeline. A nf-core standard set of labels that should be followed where possible can be seen in [nf-core/rnaseq](https://github.com/nf-core/rnaseq/blob/master/conf/base.config), which has the default process as a single core-process, and then different levels of multi-core configurations for increasingly large memory requirements defined with standardised labels.

Reviewer: ```suggestion
Sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. These should generally be specified generic with `withLabel:` selectors so they can be shared across multiple processes/steps of the pipeline. A nf-core standard set of labels that should be followed where possible can be seen in the [nf-core pipeline template](https://github.com/nf-core/tools/blob/master/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/conf/base.config), which has the default process as a single core-process, and then different levels of multi-core configurations for increasingly large memory requirements defined with standardised labels.
```

> These can be utilised within the process itself using `${task.cpu}` or `${task.memory}` variables in the `script:` block.

Reviewer: ```suggestion
The process resources can be passed on to the tool dynamically within the process with the `${task.cpu}` and `${task.memory}` variables in the `script:` block.
```

### Naming schemes

Please use the following naming schemes, to make it easy to understand what is going where.

* initial process channel: `ch_output_from_<process>`
* intermediate and terminal channels: `ch_<previousprocess>_for_<nextprocess>`

### Nextflow version bumping

> If you have agreement from reviewers, you may bump the 'default' minimum version of nextflow (e.g. for testing), with: `nf-core bump-version --nextflow . [min-nf-version]`

Reviewer: ```suggestion
If you are using a new feature from core Nextflow, you may bump the minimum required version of nextflow in the pipeline with: `nf-core bump-version --nextflow . [min-nf-version]`
```

### Software version reporting

If you add a new tool to the pipeline, please ensure you add the information of the tool to the `get_software_version` process.

Add to the script block of the process, something like the following:

```bash
<YOUR_TOOL> --version &> v_<YOUR_TOOL>.txt 2>&1 || true
```

or

```bash
<YOUR_TOOL> --help | head -n 1 &> v_<YOUR_TOOL>.txt 2>&1 || true
```

> You then need to edit the script `bin/scrape_software_versions.py` to

Reviewer: ```suggestion
You then need to edit the script `bin/scrape_software_versions.py` to:
```

> 1. add a (python) regex for your tools --version output (as in stored in the `v_<YOUR_TOOL>.txt` file), to ensure the version is reported as a `v` and the version number e.g. `v2.1.1`

Reviewer: ```suggestion
1. Add a Python regex for your tool's `--version` output (as in stored in the `v_<YOUR_TOOL>.txt` file), to ensure the version is reported as a `v` and the version number e.g. `v2.1.1`
```

> 2. add a HTML block entry to the `OrderedDict` for formatting in MultiQC.

Reviewer: ```suggestion
2. Add a HTML entry to the `OrderedDict` for formatting in MultiQC.
```

### Default processes resource requirements

> Defining sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. These should generally be specified generic with `withLabel:` selectors so they can be shared across multiple processes/steps of the pipeline. A nf-core standard set of labels that should be followed where possible can be seen in [nf-core/rnaseq](https://github.com/nf-core/rnaseq/), where the base config has the process default as a single core-process, and then different levels of multi-core jobs for increasingly large memory requirements defined with standard labels.

Reviewer 2: ```suggestion
Sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. These should generally be specified generic with `withLabel:` selectors so they can be shared across multiple processes/steps of the pipeline. A nf-core standard set of labels that should be followed where possible can be seen in [nf-core/rnaseq](https://github.com/nf-core/rnaseq/), where the base config has the process default as a single core-process, and then different levels of multi-core jobs for increasingly large memory requirements defined with standard labels.
```

> Defining sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. These should generally be specified generic with `withLabel:` selectors so they can be shared across multiple processes/steps of the pipeline. A nf-core standard set of labels that should be followed where possible can be seen in [nf-core/rnaseq](https://github.com/nf-core/rnaseq/), where the base config has the process default as a single core-process, and then different levels of multi-core jobs for increasingly large memory requirements defined with standard labels.

Reviewer 2: ```suggestion
Defining sensible defaults for process resource requirements (CPUs / memory / time) for a process should be defined in `conf/base.config`. These should generally be specified generic with `withLabel:` selectors so they can be shared across multiple processes/steps of the pipeline. A nf-core standard set of labels that should be followed where possible can be seen in [nf-core/rnaseq](https://github.com/nf-core/rnaseq/blob/master/conf/base.config), which has the default process as a single core-process, and then different levels of multi-core configurations for increasingly large memory requirements defined with standard labels.
```

Author: Oops, forgot to add this to batch commit. Will update manually (Sorry).
